ui/app/src/components/DatasourceList/DatasourceEditorForm.tsx

```diff
@@ -0,0 +1,245 @@
+import {
+  Box,
+  Button,
+  Divider,
+  FormControl,
+  FormHelperText,
+  Grid,
+  InputLabel,
+  MenuItem,
+  Select,
+  Stack,
+  TextField,
+  Typography,
+} from '@mui/material';
+import { DispatchWithoutAction, useCallback, useMemo, useState } from 'react';
+import { PluginEditor } from '@perses-dev/plugin-system';
+import { DiscardChangesConfirmationDialog } from '@perses-dev/components';
+import { useIsReadonly } from '../../model/config-client';
+
+// TODO: Replace with proper validation library
+function getValidation(state: Datasource) {
+  /** Name validation */
+  let name = null;
+  if (!state.metadata.name) {
+    name = 'Name is required';
+  }
+  // name can only contain alphanumeric characters and underscores and no spaces
+  if (state.metadata.name && !/^[a-zA-Z0-9_-]+$/.test(state.metadata.name)) {
+    name = 'Name can only contain alphanumeric characters, underscores, and dashes';
+  }
+
+  return {
+    name,
+    isValid: !name,
+  };
+}
+
+// this preprocessing ensure that we always have a defined object for the `display` property:
+function getInitialState(datasource: Datasource): Datasource {
+  const patchedDisplay: Display = {
+    name: '',
+    description: '',
+  };
+
+  if (datasource.spec.display) {
+    patchedDisplay.name = datasource.spec.display.name;
+    if (datasource.spec.display.description) {
+      patchedDisplay.description = datasource.spec.display.description;
+    }
+  }
+
+  return {
+    ...datasource,
+    spec: {
+      ...datasource.spec,
+      display: patchedDisplay,
+    },
+  };
+}
+
+interface DatasourceEditorFormProps {
+  initialDatasource: Datasource;
+  saveAction: string;
+  onSave: (datasource: Datasource) => void;
+  onClose: DispatchWithoutAction;
+  onDelete: DispatchWithoutAction | undefined;
+}
+
+export function DatasourceEditorForm(props: DatasourceEditorFormProps) {
+  const { initialDatasource, saveAction, onSave, onClose, onDelete } = props;
+  const [isDiscardDialogStateOpened, setDiscardDialogStateOpened] = useState<boolean>(false);
+  const isReadonly = useIsReadonly();
+  const initialState = getInitialState(initialDatasource);
+  const [state, setState] = useImmer(initialState);
+  const validation = useMemo(() => getValidation(state), [state]);
+
+  // When saving, remove the display property if ever display.name is empty, then pass the value upstream
+  const handleSave = () => {
+    onSave({
+      ...state,
+      spec: {
+        ...state.spec,
+        display: state.spec.display?.name !== '' ? state.spec.display : undefined,
+      },
+    });
+  };
+
+  // When the user clicks on cancel, ask for discard approval if anything was changed
+  const handleCancel = useCallback(() => {
+    if (JSON.stringify(initialState) !== JSON.stringify(state)) {
+      setDiscardDialogStateOpened(true);
+    } else {
+      onClose();
+    }
+  }, [state, initialState, setDiscardDialogStateOpened, onClose]);
+
+  return (
+    <>
+      <Box
```

Comment:
I could see some nice-to-have refactoring to break up this component in the future, also some of the spacing between form controls looks a little tight in places and slightly further apart in others:
<img width="975" alt="image" src="https://github.com/perses/perses/assets/9369625/a263d659-6ac5-4f6d-bbca-3a1f5821d658">

Reply:
Overall spacing has been improved with latest changes: 
![image](https://github.com/perses/perses/assets/7058693/cbdc0ac4-2f25-4884-ac9d-51fe26e21da5)